### PR TITLE
Add Tau leptons producer and MET Cov Matrix

### PR DESCRIPTION
**Pull request by clacaputo**
New TauProducer added to the llbb framework: `TauProducer.cc` `TauProducer.h`

Modified:   `METProducer.h` `METProducer.cc`
- MET covariance matrix added

Modified:   `Framework.py`
- new _redoMETsSignificance_ that compute MET covariance and significance
  if JEC or/and JER have been done
- taus producer added to the main sequence

### CODE REVIEW

**Comment by pieterdavid** [comment]
Thanks for the update! The tests fail with
```
An exception of category 'ProductNotFound' occurred while
   [0] Processing run: 1 lumi: 1356 event: 270939
   [1] Running path 'p'
   [2] Calling event method for module ExTreeMaker/'framework'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: ROOT::Math::SMatrix<double,2,2,ROOT::Math::MatRepSym<double,2> >
Looking for module label: METSignificance
Looking for productInstanceName: METCovariance
```
so there's still something not entirely correct (it could be as simple as adding `framework.redoMETsSignificance()` to the test configs).
I didn't have time to go through the code in detail or test yet, but from a quick look at https://github.com/cp3-llbb/Framework/pull/283/commits/b77ede81c088b869cacf8f34c7a9c5d436fd649f#diff-a340f819a1cefc5ce0a473b558712dafR378 : I would call this `addMETsSignificance` - and I think it should actually go into https://github.com/cp3-llbb/Framework/blob/03d6d79843740ceaa5fb317eb92e51a53fdd1902/python/Tools.py#L166.

**Comment by vischia** [comment]
This has still conflicts in

```
src/classes.h
src/classes_def.xml 
```

@clacaputo do you prefer to deal with them or can I try to solve them?

**Comment by clacaputo** [comment]
@vischia I will fix it and update the PR in the next days

**Comment by pieterdavid** [comment]
The only difference are the new branches, so the tests are happy :-) (you can generate new references that include the tau branches with the [`generate_reference_trees.sh`](test/generate_reference_trees.sh) script)
```
Branches found in output tree but not in the reference tree:
  met_CovMatrix00
  met_CovMatrix01
  met_CovMatrix10
  met_CovMatrix11
  met_jecdown_CovMatrix00
  met_jecdown_CovMatrix01
  met_jecdown_CovMatrix10
  met_jecdown_CovMatrix11
  met_jecup_CovMatrix00
  met_jecup_CovMatrix01
  met_jecup_CovMatrix10
  met_jecup_CovMatrix11
  met_jerdown_CovMatrix00
  met_jerdown_CovMatrix01
  met_jerdown_CovMatrix10
  met_jerdown_CovMatrix11
  met_jerup_CovMatrix00
  met_jerup_CovMatrix01
  met_jerup_CovMatrix10
  met_jerup_CovMatrix11
  tau_EAR03
  tau_EAR04
  tau_IDmap
  tau_chargedHadronIsoR03
  tau_chargedHadronIsoR04
  tau_decayMode
  tau_dz
  tau_gen_p4
  tau_gen_truth
  tau_has_matched_gen_particle
  tau_neutralHadronIsoR03
  tau_neutralHadronIsoR04
  tau_p4
  tau_photonIsoR03
  tau_photonIsoR04
  tau_puChargedHadronIsoR03
  tau_puChargedHadronIsoR04
  tau_relativeIsoR03
  tau_relativeIsoR03_deltaBeta
  tau_relativeIsoR03_withEA
  tau_relativeIsoR04
  tau_relativeIsoR04_deltaBeta
  tau_relativeIsoR04_withEA
```

File: interface/TausProducer.h [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+#include <cp3_llbb/Framework/interface/ScaleFactors.h>
+
+#include <DataFormats/VertexReco/interface/Vertex.h>
+#include <DataFormats/PatCandidates/interface/Tau.h>
+#include <DataFormats/PatCandidates/interface/TauPFSpecific.h>
+
+#include <utility>
+#include <Math/VectorUtil.h>
+
+typedef std::unordered_map<std::string, float> tauDiscriminatorMap;
```
Comment (pieterdavid) [inline interface/TausProducer.h]: Just a matter of style: could you move the typedefs in a namespace (e.g. MatchResult could be private inside TausProducer, union_sets) if they are not stored in the tree, and `Types.h` if they are stored in the output tree?
LorentzVector is defined in [`Types.h`](https://github.com/cp3-llbb/Framework/blob/f59184391fe2ee6b6cc30ecd071ba63e4fa8e242/interface/Types.h#L9) already, you can include it.

File: interface/TausProducer.h [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        edm::EDGetTokenT<std::vector<reco::Vertex>> m_vertices_token;
+	edm::EDGetTokenT<std::vector<reco::GenParticle>> m_pruned_token;
+    private:
+ 
+	MatchResult LeptonGenMatch(const LorentzVector& p4, const std::vector<reco::GenParticle>& genParticles);
+
+	inline void FindFinalStateDaughters(const reco::GenParticle& particle, std::vector<const reco::GenParticle*>& daughters,
+					    const std::set<int>& pdg_to_exclude = {})
+	{
+	  if(!particle.daughterRefVector().size()) {
```
Comment (pieterdavid) [inline interface/TausProducer.h]: detail: `if( particle.daughterRefVector().empty() )` (more efficient and more readable)

File: interface/TausProducer.h [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+	    if(excludeLightLeptons && light_leptons.count(std::abs(daughter->pdgId())) && daughter->statusFlags().isDirectTauDecayProduct()) continue;
+	    p4 += daughter->p4();
+	  }
+	  return p4;
+	}
+
+
+
+    public:
+        // Tree members
```
Comment (pieterdavid) [inline interface/TausProducer.h:126]: Do you need this commented code? If it's optional you can pass a boolean option to TausProducer, if it's not needed there's no reason to keep it around.

File: python/Framework.py [hunk 4]
```diff
@@ -372,6 +386,50 @@ def smearJets(self, resolutionFile=None, scaleFactorFile=None):
...
+                                                                  srcMet               = cms.InputTag(self.__miniaod_met_collection),
+                                                                  srcPFCandidates      = cms.InputTag('packedPFCandidates'),
+                                                                  srcJetSF             = cms.string('AK4PFchs'),
+                                                                  srcJetResPt          = cms.string('AK4PFchs_pt'),
+                                                                  srcJetResPhi         = cms.string('AK4PFchs_phi'),
+                                                                  srcRho               = cms.InputTag('fixedGridRhoAll'),
+
+                                                                  parameters = METSignificanceParams
+                                                                  )
+        # # Look for producers using the default muon input
```
Comment (pieterdavid) [inline python/Framework.py]: Same comment about commented code... should the modified MET collection be picked up by other producers/analyzers?

File: src/TausProducer.cc [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  if(!event.isRealData()){
+     edm::Handle<std::vector<reco::GenParticle>> genParticles_handle;
+     event.getByToken(m_pruned_token, genParticles_handle);
+     genParticles = *genParticles_handle;
+  }
+
+  const reco::Vertex& primary_vertex = (*vertices_handle)[0];
+
+  for (auto tau: *taus) {
+    // if (! pass_cut(tau))
```
Comment (pieterdavid) [inline src/TausProducer.cc]: commented code

File: src/TausProducer.cc [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    auto value = pair.second;
+    tauIDmap[key] = value;
+  }
+  IDmap.push_back(tauIDmap);
+
+
+  // tau decay mode
+  decayMode.push_back(tau.decayMode());
+
+  // // Same values used for cut-based tau ID. See:
```
Comment (pieterdavid) [inline src/TausProducer.cc]: commented code

**Review by pieterdavid — COMMENTED** [review]
Thanks a lot! Mostly minor and style comments.
On a different point: should these be forward-ported to 9_4_X, 10_2_X, or both? (and with which urgency? I could try cherry-picking the changes into https://github.com/cp3-llbb/Framework/pull/289)

**Comment by pieterdavid** [comment]
```
  met_CovMatrix00
  met_CovMatrix01
  met_CovMatrix10
  met_CovMatrix11
  tau_IDmap
  tau_decayMode
  tau_dz
  tau_gen_p4
  tau_gen_truth
  tau_has_matched_gen_particle
  tau_p4
```
with jet systematic variations for the met covariance. Nice :-)

**Comment by clacaputo** [comment]
```
All tests passed (20087 assertions in 1 test case)

status = 256

---> test Testcp3_llbbFramework had ERRORS
 
^^^^ End Test Testcp3_llbbFramework ^^^^
>> Tests for package cp3_llbb/Framework ran.
+ popd
~/workspace/Framework-SLC6/CMSSW/CMSSW_8_0_30/src ~/workspace/Framework-SLC6
+ grep ERRORS ../logs/slc6_amd64_gcc530/testing.log
---> test Testcp3_llbbFramework had ERRORS
Build step 'Execute shell' marked build as failure
Setting status of 79e4f88895abff3e28b68d996e926a16ddda3293 to FAILURE with url https://jenkins-cp3-llbb.web.cern.ch/job/Framework-SLC6/463/ and message: 'Build finished. '
Finished: FAILURE
```

**Comment by pieterdavid** [comment]
The error is because you forgot to update the data reference file

**Pull request merged**